cassiopeia/tools/branch_length_estimator/IIDExponentialPosteriorMeanBLE.py

```diff
@@ -0,0 +1,537 @@
+        # Pull out params
+        r = self.mutation_rate
+        lam = self.birth_rate
+        dt = 1.0 / self.discretization_level
+        K = self.tree.num_characters()
+        tree = self.tree
+        assert v != tree.root()
+        assert 0 <= t <= self.discretization_level
+        assert 0 <= x <= K
+        if not (1.0 - lam * dt - K * r * dt > 0):
+            raise ValueError("Please choose a bigger discretization_level.")
+        log_likelihood = 0.0
+        if t == 0:  # Base case
+            if v in tree.leaves() and x == tree.num_cuts(v):
+                # TODO: 'v not in tree.leaves()' is O(n). We should have O(1)
+                # check.
+                log_likelihood = 0.0
+            else:
+                log_likelihood = -np.inf
+        else:  # Recursion.
+            log_likelihoods_cases = []
+            # Case 1: Nothing happens
+            log_likelihoods_cases.append(
+                np.log(1.0 - lam * dt - (K - x) * r * dt)
+                + self.down(v, t - 1, x)
+            )
+            # Case 2: One character mutates.
+            if x + 1 <= K:
+                log_likelihoods_cases.append(
+                    np.log((K - x) * r * dt) + self.down(v, t - 1, x + 1)
+                )
+            # Case 3: Cell divides
+            # The number of cuts at this state must match the ground truth.
+            # TODO: Allow for weak match at internal nodes and exact match at
+            # leaves.
+            if (
+                self.compatible_with_observed_data(x, tree.num_cuts(v))
+                and v not in tree.leaves()
+            ):
+                # TODO: 'v not in tree.leaves()' is O(n). We should have O(1)
+                # check.
+                ll = sum(
+                    [self.down(child, t - 1, x) for child in tree.children(v)]
+                ) + np.log(lam * dt)
+                log_likelihoods_cases.append(ll)
+            log_likelihood = logsumexp(log_likelihoods_cases)
+        self.down_cache[(v, t, x)] = log_likelihood
+        return log_likelihood
+
+    @classmethod
+    def exact_log_full_joint(
+        self, tree: Tree, mutation_rate: float, birth_rate: float
+    ) -> float:
+        r"""
+        log P(T, X, branch_lengths), i.e. the full joint log likelihood given
+        both character vectors _and_ branch lengths.
+        """
+        tree = deepcopy(tree)
+        tree.set_edge_lengths_from_node_ages()
+        ll = 0.0
+        lam = birth_rate
+        r = mutation_rate
+        lg = np.log
+        e = np.exp
+        b = binom
+        for (p, c) in tree.edges():
+            t = tree.get_edge_length(p, c)
+            # Birth process likelihood
+            ll += -t * lam
+            if c not in tree.leaves():
+                ll += lg(lam)
+            # Mutation process likelihood
+            cuts = tree.number_of_mutations_along_edge(p, c)
+            uncuts = tree.number_of_nonmutations_along_edge(p, c)
```

Comment:
I might be wrong, but given your implementation of the `number_of_nonmutations_along_edge` in the `Tree` class, you might be underestimating the number of sites that don't mutate. Can you double check this is doing what you want?